ts_src/silentpayment.ts

```diff
@@ -0,0 +1,282 @@
+      throw new Error(
+        'Invalid public key length, expected 33 bytes public key',
+      );
+    }
+  }
+
+  static decode(str: string): SilentPaymentAddress {
+    const result = bech32m.decode(str, 118);
+    const version = result.words.shift();
+    if (version !== 0) {
+      throw new Error('Unexpected version of silent payment code');
+    }
+    const data = bech32m.fromWords(result.words);
+    const scanPubKey = Buffer.from(data.slice(0, 33));
+    const spendPubKey = Buffer.from(data.slice(33));
+    return new SilentPaymentAddress(spendPubKey, scanPubKey);
+  }
+
+  encode(): string {
+    const data = Buffer.concat([this.scanPublicKey, this.spendPublicKey]);
+
+    const words = bech32m.toWords(data);
+    words.unshift(0);
+    return bech32m.encode('sp', words, 118);
+  }
+}
+
+// inject ecc dependency, returns a SilentPayment interface
+export function SPFactory(ecc: TinySecp256k1Interface): SilentPayment {
+  return new SilentPaymentImpl(ecc);
+}
+
+const SEGWIT_V1_SCRIPT_PREFIX = Buffer.from([0x51, 0x20]);
+
+const G = Buffer.from(
+  '0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798',
+  'hex',
+);
+
+class SilentPaymentImpl implements SilentPayment {
+  constructor(private ecc: TinySecp256k1Interface) {}
+
+  /**
+   * Compute scriptPubKey used to send funds to a silent payment address
+   * @param inputs list of ALL outpoints of the transaction sending to the silent payment address
+   * @param inputPrivateKey private key owning the spent outpoint. Sum of all private keys if multiple inputs
+   * @param silentPaymentAddress target of the scriptPubKey
+   * @param index index of the silent payment address. Prevent address reuse if multiple silent addresses are in the same transaction.
+   * @returns the output scriptPubKey belonging to the silent payment address
+   */
+  makeScriptPubKey(
+    inputs: Outpoint[],
+    inputPrivateKey: Buffer,
+    silentPaymentAddress: string,
+    index = 0,
+  ): Buffer {
+    const inputsHash = hashOutpoints(inputs);
+    const addr = SilentPaymentAddress.decode(silentPaymentAddress);
+
+    const sharedSecret = this.makeSharedSecret(
+      inputsHash,
+      addr.scanPublicKey,
+      inputPrivateKey,
+    );
+
+    const outputPublicKey = this.makePublicKey(
+      addr.spendPublicKey,
+      index,
+      sharedSecret,
+    );
+
+    return Buffer.concat([SEGWIT_V1_SCRIPT_PREFIX, outputPublicKey.slice(1)]);
+  }
+
+  /**
+   * Check if a scriptPubKey belongs to a silent payment address
+   * @param scriptPubKey scriptPubKey to check
+   * @param inputs list of ALL outpoints of the transaction sending to the silent payment address
+   * @param inputPublicKey public key owning the spent outpoint. Sum of all public keys if multiple inputs
+   * @param scanSecretKey *scan* secret key of the silent payment address
+   * @param spendPublicKey *spend* public key of the silent payment address
+   * @param index index of the silent payment address.
+   */
+  isMine(
```

Comment:
same for this "wrapper", in the end it's only a Buffer.compare.. would be better to implement this kind of "high-level" functions into the projects using liquidjs as a dep.